### PR TITLE
ALC-301 - calculate liquidity value based on token balances

### PR DESCRIPTION
**Pull request by hosainnet**
For MVP. Uses logic based on https://github.com/balancer-labs/pool-management-vue/blob/develop/src/helpers/price.ts#L3-L21 to calculate the pool liquidity based on the current token prices and token balances in the pool

Example: https://pools.balancer.exchange/#/pool/0xf823905b5aca1e810cc458dc013598552e1682ae/

Before:

Screenshot 2021-05-03 at 18 47 59

After:

Screenshot 2021-05-03 at 18 50 51

The difference is down to the source of token prices

- [x] Branch name corresponds to Linear ticket number.
- [x] Added sufficient integration and unit tests.
- [x] Addressed a single concern in the least number of changed lines as
      possible.
- [x] Only fixed/added the functionality in question **OR** address wide-spread
      whitespace/style issues, not both.

### CODE REVIEW

File: src/components/PoolsList/__mock__/graphql-responses.ts [hunk 1]
```diff
@@ -48,4 +52,33 @@ export const POOL_QUERY_RESPONSE = [
...
+          },
+          {
+            id: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
+            price: "3169.613887121700198292120988763966",
+            symbol: "WETH",
+          },
+        ],
+      },
+    },
+  },
```
Comment (hosainnet) [inline src/components/PoolsList/__mock__/graphql-responses.ts:83]: Update graphql mocks the dashboard

File: src/components/PoolsList/__mock__/graphql-responses.ts [hunk 2]
```diff
@@ -34,6 +37,7 @@ export const POOL_QUERY_RESPONSE = [
                   "0x002ad19fb25c6206d6d19e524f363ea846afe4a5-0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                 name: "Wrapped Ether",
                 symbol: "WETH",
+                balance: 25,
```
Comment (hosainnet) [inline src/components/PoolsList/__mock__/graphql-responses.ts:40]: Used for liquidity pool later

File: src/pages/PoolDashboard/index.tsx [hunk 3]
```diff
@@ -23,6 +25,11 @@ const PoolDashboard = ({ poolDetails }: PoolDashboardProps) => {
   const { loading, error, data } = useQuery(
     getPoolMetrics(poolId, daysFromStart)
   );
+  const {
+    loading: loadingPricesLoading,
+    error: poolPricesError,
+    data: tokenPricesResponse,
+  } = useQuery(getTokenPrices());
```
Comment (hosainnet) [inline src/pages/PoolDashboard/index.tsx:32]: Query prices to be used for liquidity calculations

File: src/queries/pools.ts [hunk 4]
```diff
@@ -57,6 +57,7 @@ export const GET_VALID_POOL_BY_ID = gql`
         symbol
         address
         denormWeight
+        balance
```
Comment (hosainnet) [inline src/queries/pools.ts:60]: Pull the required field

File: src/queries/prices.ts [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
-import constants from "../constants";
 
 export const getTokenPrices = () => {
   let query = {
     tokenPrices: {
       __args: {
+        first: 1000,
         where: {
-          id_in: Object.keys(constants.TOKEN_ADDRESS_TO_SYMBOL),
+          price_gt: 0,
```
Comment (hosainnet) [inline src/queries/prices.ts:10]: Bring back all token prices, this query is now similar to https://github.com/balancer-labs/pool-management-vue/blob/develop/src/helpers/queries.json#L15-L31

File: src/utils/mocks/pools.ts [hunk 6]
```diff
@@ -87,3 +87,39 @@ export const onlyPokeWeightTransactions = [
...
+  swapFee: 1,
+  totalSwapFee: 100,
+  controller: "0x8079aadf0514e6b2d6ffc19a5a42c95ee83cabae",
+  publicSwap: true,
+  finalized: false,
+  liquidity: 993807.6803026758236516899810813218,
+  totalSwapVolume: 3764736.790516105364859220839915875,
+  totalWeight: 25,
+  createTime: 1615812993,
+};
```
Comment (hosainnet) [inline src/utils/mocks/pools.ts:125]: Moved from one of the unit tests so it can be re-used

File: src/queries/prices.ts (hunk 5 above)
Comment (risk-1) [reply src/queries/prices.ts:10]: hm. this brings back the first 1000? what's our guarantee that the token we care about is captured by it? unless we know balancer has a lot less tokens, and we change this approach post MVP

Comment (risk-1) [reply src/queries/prices.ts:10]: also why not configure this to only fetch the prices of the two currencies we care about about based on the pool info? just the fund and launch tokens. currently the call to fetch all the prices takes a really long time (5 seconds).

File: src/utils/price.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+export function mapPrices(tokenPrices: TokenPriceEntry[]): TokenToPrice {
+  return Object.fromEntries(
+    tokenPrices.map((tokenPrice) => [
+      getAddress(tokenPrice.id),
+      parseFloat(tokenPrice.price),
+    ])
+  );
+}
+
+export function calculatePoolLiquidity(
```
Comment (risk-1) [inline src/utils/price.ts:15]: there seems to be a tiny mismatch on the numbers but could be due to some sort of rounding issues. Can file a bug for now and revisit post mvp.
Screen Shot 2021-05-03 at 6 21 30 PM
Screen Shot 2021-05-03 at 6 20 52 PM

File: src/pages/PoolDashboard/index.tsx [hunk 8]
```diff
@@ -54,6 +61,14 @@ const PoolDashboard = ({ poolDetails }: PoolDashboardProps) => {
       });
   }, []);
 
+  if (loading || loadingPricesLoading) {
+    return <p>Loading...</p>;
+  }
```
Comment (ledes) [inline src/pages/PoolDashboard/index.tsx:66]: nice! we eventually need to consider every call to render a loading state (getPoolTransactions too)
For the ones that don't use useQuery we can just have a boolean in the state of the page. We can do this whenever we add a loading component

File: src/queries/prices.ts [hunk 9]
```diff
@@ -1,13 +1,13 @@
 import { gql } from "@apollo/client/core";
 import { jsonToGraphQLQuery } from "json-to-graphql-query";
-import constants from "../constants";
 
 export const getTokenPrices = () => {
   let query = {
     tokenPrices: {
       __args: {
+        first: 1000,
```
Comment (ledes) [inline src/queries/prices.ts:8]: why 1000?

File: src/utils/price.ts (hunk 7 above)
Comment (ledes) [inline src/utils/price.ts:15]: this is sick!

**Review by risk-1 — APPROVED** [review]

**Comment by risk-1** [comment]
approved given that we wont pull full price list and just constrain it to the two currencies

File: src/queries/prices.ts (hunk 9 above)
Comment (hosainnet) [reply src/queries/prices.ts]: Removed, the idea was to cache those so moving around the site is quick but for MVP I've changed it to load just the pool's price tokens

File: src/pages/PoolDashboard/index.tsx (hunk 8 above)
Comment (hosainnet) [reply src/pages/PoolDashboard/index.tsx:68]: 👍

File: src/queries/prices.ts (hunk 5 above)
Comment (hosainnet) [reply src/queries/prices.ts]: @risk-1 changed, give it another go please to see if you notice the difference

**Review by ledes — APPROVED** [review]

**Pull request merged**